src/tapedfunction.jl

```diff
@@ -138,124 +162,97 @@ function (instr::Instruction{typeof(_new)})()
-    elseif arity == 1
-        args_var = IRTools.insertafter!(ir, tape, IRTools.xcall(@__MODULE__, :box_args, args[2]))
-        arg_pairs = Dict(args[2] => args_var)
-    else # arity > 1
-        args_var = IRTools.insertafter!(ir, tape, IRTools.xcall(@__MODULE__, :box_args, args[2:end]...))
-        args_new, last_pos = [], args_var
 
-        iter_state = []
+## internal functions
 
-        for i in 1:arity
-            last_pos = IRTools.insertafter!(ir, last_pos, IRTools.xcall(Base, :indexed_iterate, args_var, i, iter_state...))
-            args_iter = last_pos
-            last_pos = IRTools.insertafter!(ir, last_pos, IRTools.xcall(Core, :getfield, args_iter, 1))
-            push!(args_new, last_pos)
-            if i != arity
-                last_pos = IRTools.insertafter!(ir, last_pos, IRTools.xcall(Core, :getfield, args_iter, 2))
-                iter_state = [last_pos]
-            end
+arg_boxer(var, boxes) = var
+arg_boxer(var::IRTools.Variable, boxes) = get!(boxes, var, Box{Any}(nothing))
+function args_initializer(ins::BlockInstruction)
+    return (args...) -> begin
+        @assert length(args) + 1 == length(ins.args)
+        ins.args[1].val = ins.tape.func
+        for i in 1:length(args)
+            ins.args[i + 1].val = val(args[i]) # fill the boxes
         end
-        arg_pairs = Dict(zip(args[2:end], args_new))
     end
+end
 
-    # here we assumed the ir only has a return statement at its last block,
-    # and we make sure the return value is from a function call (to `identity`)
-    last_blk = IRTools.blocks(ir)[end]
-    retv = IRTools.returnvalue(last_blk)
-    IRTools.return!(last_blk, IRTools.xcall(Base, :identity, retv))
+function translate!(taped::Taped, ir::IRTools.IR)
+    tape = taped.tape
+    boxes = Dict{IRTools.Variable, Box{Any}}()
+    _box = (x) -> arg_boxer(x, boxes)
+    for (blk_id, blk) in enumerate(IRTools.blocks(ir))
+        # blocks
+        blk_args = IRTools.arguments(blk)
+        push!(tape, BlockInstruction(blk_id, map(_box, blk_args), taped))
+        # `+ 1` because we will have an extra ins at the beginning
+        taped.block_map[blk_id] = length(tape) + 1
+        # normal instructions
+        for (x, st) in blk
+            if Meta.isexpr(st.expr, :call)
+                args = map(_box, st.expr.args)
+                # args[1] is the function
+                f = args[1]
+                ins = Instruction(f, args[2:end] |> Tuple,
+                                  _box(x), taped)
+                push!(tape, ins)
+            elseif Meta.isexpr(st.expr, :new)
+                args = map(_box, st.expr.args)
+                ins = Instruction(_new, args |> Tuple, _box(x), taped)
+                push!(tape, ins)
+            elseif isa(st.expr, GlobalRef)
+                ins = Instruction(val, (st.expr,), _box(x), taped)
+                push!(tape, ins)
+            else
+                @warn "Unknown IR code: " st
```

Comment:
I think we should error here so that user knows something went wrong. 
```suggestion
                @error "Unknown IR code: " st
```